lib/actions/api.js

```diff
@@ -130,6 +165,216 @@ export function bikeRentalQuery () {
+
+// Single Route lookup query
+
+export const findRouteResponse = createAction('FIND_ROUTE_RESPONSE')
+export const findRouteError = createAction('FIND_ROUTE_ERROR')
+
+export function findRoute (params) {
+  return createQueryAction(`index/routes/${params.routeId}`,
+    findRouteResponse, findRouteError, null,
+    (payload, dispatch) => {
+      // load patterns
+      dispatch(findPatternsForRoute({ routeId: params.routeId }))
+    }
+  )
+}
+
+// Patterns for Route lookup query
+
+export const findPatternsForRouteResponse = createAction('FIND_PATTERNS_FOR_ROUTE_RESPONSE')
+export const findPatternsForRouteError = createAction('FIND_PATTERNS_FOR_ROUTE_ERROR')
+
+export function findPatternsForRoute (params) {
+  return createQueryAction(`index/routes/${params.routeId}/patterns`,
+    findPatternsForRouteResponse, findPatternsForRouteError,
+    (payload) => {
+      // convert pattern array to ID-mapped object
+      const patterns = {}
+      payload.forEach(ptn => { patterns[ptn.id] = ptn })
+
+      return {
+        routeId: params.routeId,
+        patterns
+      }
+    },
+    (payload, dispatch) => {
+      // load geometry for each pattern
+      payload.forEach(ptn => {
+        dispatch(findGeometryForPattern({
+          routeId: params.routeId,
+          patternId: ptn.id
+        }))
+      })
+    }
+  )
+}
+
+// Geometry for Pattern lookup query
+
+export const findGeometryForPatternResponse = createAction('FIND_GEOMETRY_FOR_PATTERN_RESPONSE')
+export const findGeometryForPatternError = createAction('FIND_GEOMETRY_FOR_PATTERN_ERROR')
+
+export function findGeometryForPattern (params) {
+  return createQueryAction(`index/patterns/${params.patternId}/geometry`,
+    findGeometryForPatternResponse, findGeometryForPatternError,
+    (payload) => {
+      return {
+        routeId: params.routeId,
+        patternId: params.patternId,
+        geometry: payload
+      }
+    }
+  )
+}
+
+// generic helper for constructing API queries
+
+function createQueryAction (endpoint, responseAction, errorAction, rewritePayload, postprocess) {
+  return async function (dispatch, getState) {
+    const otpState = getState().otp
+    const api = otpState.config.api
+    const url = `${api.host}${api.port ? ':' + api.port : ''}${api.path}/${endpoint}`
+    let payload
+    try {
+      const response = await fetch(url)
+      if (response.status >= 400) {
+        const error = new Error('Received error from server')
+        error.response = response
+        throw error
+      }
+      payload = await response.json()
+    } catch (err) {
+      return dispatch(errorAction(err))
+    }
+
+    if (typeof rewritePayload === 'function') {
+      dispatch(responseAction(rewritePayload(payload)))
+    } else {
+      dispatch(responseAction(payload))
+    }
+
+    if (typeof postprocess === 'function') {
+      postprocess(payload, dispatch, getState)
+    }
+  }
+}
+
+// TODO: convert the rest of these to use the above createQueryAction helper
```

Comment:
We may even want to try to convert these to use [woonerf fetch](https://github.com/conveyal/woonerf#fetch).  Maybe once we get around to refactoring we can revisit and standardize all of our api calls.

Reply:
I have replaced the remaining API calls with `createQueryAction`

Let's revisit woonerf#fetch in #45.